File: grpc-java/grpc-java-app/src/test/java/com/example/StreamingJavaTest.java
package com.example;

import com.example.foo.bar.Greeter2Grpc;
import com.example.foo.bar.HelloReply;
import com.example.foo.bar.HelloRequest;
import com.example.interceptors.Headers;
import com.example.services.ClientService;
import com.example.services.ServerService;
import com.example.utils.GrpcUtils;
import io.grpc.stub.StreamObserver;
import org.junit.jupiter.api.Test;

import static com.example.utils.LoggingUtils.log;
import static com.example.utils.TestUtils.MSG1;
import static com.example.utils.TestUtils.MSG2;
import static com.example.utils.TestUtils.MSG3;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;

class StreamingJavaTest {
    @Test
    void test() throws Exception {
        var serverService = mock(ServerService.class);
        var clientService = mock(ClientService.class);

        try (var ignored = GrpcUtils.startServer(new ServerImpl(serverService))) {
            try (var channel = GrpcUtils.createChannel()) {
                var stub = GrpcUtils.createStub(channel.channel(), Greeter2Grpc::newStub);
                var requestObserver = stub.sayHello(
                    new StreamObserver<>() {
                        @Override
                        public void onNext(HelloReply reply) {
                            log("newStub.sayHello %s", reply);
                            clientService.call(reply.getMsg());
                        }

                        @Override
                        public void onError(Throwable throwable) {
                            log("newStub.onError %s", throwable.getMessage());
                        }

                        @Override
                        public void onCompleted() {
                            log("newStub.onCompleted");
                        }
                    }
                );

                requestObserver.onNext(HelloRequest.newBuilder().setMsg(MSG1).build());
                requestObserver.onNext(HelloRequest.newBuilder().setMsg(MSG2).build());
                requestObserver.onNext(HelloRequest.newBuilder().setMsg(MSG3).build());
                requestObserver.onCompleted();
            }
        }

        var serverInOrder = inOrder(serverService);
        serverInOrder.verify(serverService).call(MSG1);
        serverInOrder.verify(serverService).call(MSG2);
        serverInOrder.verify(serverService).call(MSG3);
        serverInOrder.verifyNoMoreInteractions();

        var clientInOrder = inOrder(clientService);
        clientInOrder.verify(clientService).call(MSG1.repeat(3));
        clientInOrder.verify(clientService).call(MSG2.repeat(3));
        clientInOrder.verify(clientService).call(MSG3.repeat(3));
        clientInOrder.verifyNoMoreInteractions();
    }

    private static class ServerImpl extends Greeter2Grpc.Greeter2ImplBase {
        private final ServerService service;

        public ServerImpl(ServerService service) {
            this.service = service;
        }

        @Override
        public StreamObserver<HelloRequest> sayHello(StreamObserver<HelloReply> responseObserver) {
            return new StreamObserver<>() {
                @Override
                public void onNext(HelloRequest request) {
                    String requestId = Headers.REQUEST_ID_CTX_KEY.get();
                    log("[%s] ServerImpl.onNext %s", requestId, request);
                    service.call(request.getMsg());
                    var reply = HelloReply.newBuilder().setMsg(request.getMsg().repeat(3)).build();
                    responseObserver.onNext(reply);
                }

                @Override
                public void onError(Throwable throwable) {
                    log("ServerImpl.onError %s%n", throwable.getMessage());
                    responseObserver.onError(throwable);
                }

                @Override
                public void onCompleted() {
                    log("ServerImpl.onCompleted");
                    responseObserver.onCompleted();
                }
            };
        }
    }
}

File: demo-grpc-proto/src/test/kotlin/com/example/greeter2/StreamingJavaTest.kt
package com.example.greeter2

import com.example.util.*
import foo.bar.Greeter.HelloReply
import foo.bar.Greeter.HelloRequest
import foo.bar.Greeter2Grpc
import io.grpc.ClientInterceptors
import io.grpc.Grpc
import io.grpc.ServerInterceptors
import io.grpc.stub.StreamObserver
import org.junit.jupiter.api.Test
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.mock
import org.mockito.kotlin.timeout

private val PORT = findFreePort()

class StreamingJavaTest {
    @Test
    fun test() {
        val serverService = mock<ServerService>()
        val server = Grpc.newServerBuilderForPort(PORT, serverCredentials(false))
            .addService(ServerInterceptors.intercept(ServerImpl(serverService), HeaderServerInterceptor()))
            .build()
            .start()
        println("Server started on port: $PORT")

        val clientService = mock<ClientService>()
        val channel = Grpc.newChannelBuilder("localhost:$PORT", clientCredentials(false)).build()
        val stub = Greeter2Grpc.newStub(ClientInterceptors.intercept(channel, HeaderClientInterceptor()))
        val requestObserver = stub.sayHello(
            object : StreamObserver<HelloReply> {
                override fun onNext(reply: HelloReply) {
                    println("newStub.sayHello $reply".trim() + " thread: ${Thread.currentThread()}")
                    clientService.call(reply.msg)
                }

                override fun onError(th: Throwable) = println("newStub.onError ${th.stackTraceToString()}")
                override fun onCompleted() = println("newStub.onCompleted")
            })

        requestObserver.onNext(HelloRequest.newBuilder().setMsg("Abc").build())
        requestObserver.onNext(HelloRequest.newBuilder().setMsg("Xyz").build())
        requestObserver.onNext(HelloRequest.newBuilder().setMsg("Klm").build())
        requestObserver.onCompleted()

        val serverInOrder = inOrder(serverService)
        serverInOrder.verify(serverService, timeout(5000)).call("Abc")
        serverInOrder.verify(serverService, timeout(5000)).call("Xyz")
        serverInOrder.verify(serverService, timeout(5000)).call("Klm")
        serverInOrder.verifyNoMoreInteractions()

        val clientInOrder = inOrder(clientService)
        clientInOrder.verify(clientService, timeout(5000)).call("Abc".repeat(3))
        clientInOrder.verify(clientService, timeout(5000)).call("Xyz".repeat(3))
        clientInOrder.verify(clientService, timeout(5000)).call("Klm".repeat(3))
        clientInOrder.verifyNoMoreInteractions()

        server.shutdown()
        server.awaitTermination()
    }
}

private class ServerImpl(private val service: ServerService) : Greeter2Grpc.Greeter2ImplBase() {
    override fun sayHello(responseObserver: StreamObserver<HelloReply>): StreamObserver<HelloRequest> {
        return object : StreamObserver<HelloRequest> {
            override fun onNext(request: HelloRequest) {
                println("ServerImpl.onNext $request".trim() + " thread: ${Thread.currentThread()}")
                service.call(request.msg)
                val reply = HelloReply.newBuilder().setMsg(request.msg.repeat(3)).build()
                responseObserver.onNext(reply)
            }

            override fun onError(th: Throwable) {
                println("ServerImpl.onError ${th.stackTraceToString()}")
                responseObserver.onError(th)
            }

            override fun onCompleted() {
                println("ServerImpl.onCompleted")
                responseObserver.onCompleted()
            }
        }
    }
}


